framework/Floats.c

```diff
@@ -4,82 +4,114 @@
+static_assert(FLOAT32_INFINITY_EXPONENT == FLOAT32_EXPONENT_MASK, "Float32::Inf has all exponent bits set");
+static_assert(MASK(FLOAT32_EXPONENT_BITS) == FLOAT32_EXPONENT_MASK, "Float32 exponent mask has different size than the field");
+static_assert(MASK(FLOAT32_MANTISSA_BITS) == FLOAT32_MANTISSA_MASK, "Float32 mantissa mask has different size than the field");
+static_assert(QUIET(FLOAT32) == FLOAT32_MANTISSA_QUIET_NAN_MASK, "Quiet bit is MSB of the mantissa");
+static_assert(FLOAT32_SIGN_BITS + FLOAT32_EXPONENT_BITS + FLOAT32_MANTISSA_BITS == 32, "Bitfields sums to type size");
+
+static_assert(sizeof(Float64) == sizeof(double), "Float64 structure is not of the correct size");
+static_assert(FLOAT64_NAN_EXPONENT == FLOAT64_EXPONENT_MASK, "Float64::NaNs have all exponent bits set");
+static_assert(FLOAT64_INFINITY_EXPONENT == FLOAT64_EXPONENT_MASK, "Float64::Inf has all exponent bits set");
+static_assert(MASK(FLOAT64_EXPONENT_BITS) == FLOAT64_EXPONENT_MASK, "Float64 exponent mask has different size than the field");
+static_assert(MASK(FLOAT64_MANTISSA_BITS) == FLOAT64_MANTISSA_MASK, "Float64 mantissa mask has different size than the field");
+static_assert(QUIET(FLOAT64) == FLOAT64_MANTISSA_QUIET_NAN_MASK, "Quiet bit is MSB of the mantissa");
+static_assert(FLOAT64_SIGN_BITS + FLOAT64_EXPONENT_BITS + FLOAT64_MANTISSA_BITS == 64, "Bitfields sums to type size");
+
+static_assert(sizeof(double) < sizeof(long double), "Compiler does not support long double");
+static_assert(sizeof(Float80) == sizeof(long double), "Float80 structure is not of the correct size");
+static_assert(FLOAT80_NAN_EXPONENT == FLOAT80_EXPONENT_MASK, "All Float80::NaNs have all exponent bits set");
+static_assert(FLOAT80_INFINITY_EXPONENT == FLOAT80_EXPONENT_MASK, "Float80::Inf has all exponent bits set");
+static_assert(MASK(FLOAT80_EXPONENT_BITS) == FLOAT80_EXPONENT_MASK, "Float80 exponent mask has different size than the field");
+static_assert(MASK(FLOAT80_MANTISSA_BITS) == FLOAT80_MANTISSA_MASK, "Float80 mask has different size than mantissa+jbit");
+static_assert(QUIET(FLOAT80) == FLOAT80_MANTISSA_QUIET_NAN_MASK, "Quiet bit is MSB of the mantissa");
+static_assert(FLOAT80_SIGN_BITS + FLOAT80_EXPONENT_BITS + FLOAT80_JBIT_BITS + FLOAT80_MANTISSA_BITS == 80, "Bitfields sums to type size");
+
+
+// C++ validators
+#ifdef __cplusplus
+static const Float16 Float16_all_ones = new_float16(~0, ~0, ~0);
+
+static_assert(Float16_all_ones.sign == 1, "Float16: Incorrect sign");
+static_assert(Float16_all_ones.exponent == FLOAT16_NAN_EXPONENT, "Float16: Incorrect exponent");
+static_assert(Float16_all_ones.mantissa == FLOAT16_MANTISSA_MASK, "Float16: Incorrect mantissa");
+
+#endif
+
+
+/** @} */
+
+// function definitions
+Float16 new_random_float16()
 {
     Float16 f;
-    f.sign = sign & 1u;
-    f.exponent = exponent & FLOAT16_EXPONENT_MASK;
-    f.mantissa = mantissa & FLOAT16_MANTISSA_MASK;
+    f.as_hex = random();
```

Comment:
Getting all the bits at random has a problem of getting mantissa == 0, which earlier implementation never returned.
I am not sure if we can do that. Some implementations might depend on that.
Consider reading `random32()` return value to local variable and use the bits from a single call to fill the parts of the float value. 

Something like this:
```C
    uint32_t random_bits = random32();
    
    Float32 f;
    
    f.sign = random_bits & 1u;
    random_bits >>= 1;
    
    f.exponent = random_bits & FLOAT32_EXPONENT_MASK;
    random_bits >>= FLOAT16_EXPONENT_BITS;
    
    // This internally calls `random()`, but maybe we can change it and pass random_bits as a random bits generator/pool
    f.mantissa = set_random_bits((random_bits % 9) + 1, 10);
```